### PR TITLE
[FLINK-33915][ci] Adds nightly workflow for the master branch

### PR DESCRIPTION
**Pull request by XComp**
Based on the following PR(s):
* https://github.com/apache/flink/pull/23961
* https://github.com/apache/flink/pull/23962
* https://github.com/apache/flink/pull/23965
* https://github.com/apache/flink/pull/24061
* https://github.com/apache/flink/pull/23964
* https://github.com/apache/flink/pull/23970
* === THIS PR ===
* https://github.com/apache/flink/pull/23972

## What is the purpose of the change

Creates an extended CI workflow that runs on master

## Brief change log

* Adds custom actions for selecting the release branch and commit hash
* Adds custom action for selecting the workflow configuration
* Adds fallback workflow configuration
* Adds workflow for utilizing the workflow configuration

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: .github/actions/select_workflow_configs/action.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    value: "${{ steps.workflow-selection.outputs.configs }}"
+runs:
+  using: "composite"
+  steps:
+    - name: "Loads Flink workflow configurations"
+      id: workflow-selection
+      shell: bash
+      run: |
+        fallback_profile="default"
+        fallback_path="${{ inputs.config-folder }}/flink-workflow.${fallback_profile}.json"
```
Comment (zentol) [inline .github/actions/select_workflow_configs/action.yml:45]: why have a fallback path? shouldnt we fail hard if the parameter is specified incorrectly?

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml:45]: ...to enable nightly build runs through workflow-dispatch on branches.

Comment (zentol) [reply .github/actions/select_workflow_configs/action.yml:45]: > Why can't we use the configuration from the respective branch? checkout branch, use nightly config.

I don't remember how exactly GHA handles things w.r.t. which branches workflow files are being honored...

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml:45]: > a) nightly builds should use workflow_call; workflow_dispatch is for manual runs via the UI

I'm not sure I understand your statement here. The default json file is essentially the workflow for the nightly build of master. Additionally, I enabled `workflow_dispatch` as an valid event trigger for this workflow as well so that anyone could trigger an extended workflow run (with the same configuration as the nightly build on `master`) on his/her branch.

> shouldn't master explicitly point to the default config?

Fine with me. We could rename the to `flink-workflow.master.json` and still use it as fallback for branches that do not have a dedicated configuration. If that's what you suggest.

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml:45]: > Why can't we use the configuration from the respective branch? checkout branch, use nightly config.

Scheduled workflow runs will be triggered from `master`. The workflow configuration (the json file) is therefore also loaded from `master`. The workflow then triggers the checkout of individual branches.

> Do we plan to have different workflow configs for a single branch?

We have that right now: 1.18 doesn't include Java 21. But you're right: It's not a common case, I'd say.

There are some reasons why I came up with this solution (having dedicated per-branch configuration files and one workflow file per branch):

Initially, I went for a single workflow yaml that would trigger a fixed list of release branches through the matrix strategy.  We could have disabled the Java 21 workflow. But the flaw of this approach is that the workflow will be nested in three levels:
1. branch
2. profiles
3. the individual jobs of flink-ci.template.yml

GitHub Actions doesn't work well with three levels. The dropdown per profile would go away. Instead, we would have a dropdown per branch with all profiles and their individual jobs flattened out in a long list. Unfortunately, I don't have a workflow run saved anymore which would show what I mean. But It's essentially similar what we have with Azure Pipelines: A long list of individual jobs that are hard to parse by the human eye. 

Having individual workflow yamls for each branch helped work around this issue with the cost of having kind of redundant code. I introduced the workflow configuration JSON files to reduce the redundancy if the workflow configuration to a minimum.

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml:45]: > Scheduled workflow runs will be triggered from master. The workflow configuration (the json file) is therefore also loaded from master. The workflow then triggers the checkout of individual branches.

The following build failure (which happened in release-1.18) seems to prove me wrong: https://github.com/XComp/flink/actions/runs/7498584099

Indeed, dedicated config files (e.g. custom actions and as a consequence also loaded from the branch). So your solution is viable. My reasoning that all the config is picked up from master while scheduling the build appears to be wrong. That means that we could have one workflow configuration file per branch. We will still keep the functionality around configuration files in general to avoid having too much duplicate code for each of the nightly workflows.

Comment (zentol) [reply .github/actions/select_workflow_configs/action.yml:45]: > Scheduled workflow runs will be triggered from master. The workflow configuration (the json file) is therefore also loaded from master. The workflow then triggers the checkout of individual branches.

IIRC the only thing that comes from master is the main workflow file, but everything else comes from the branch, I think because we use relative paths to links to the composite actions.

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml]: hm, looks like we are wrong: A [test workflow run](https://github.com/XComp/flink/actions/runs/7504580580/job/20431989469) on my fork's release-1.18 branch (where I had all the changes backported) picked up the Java 21 configuration (it's based on https://github.com/apache/flink/commit/274292a1e275e1202280748d13592ee9e165e121 which included the removal of the Java 21 sub-workflow). That's weird.

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml]: Which, in the end makes sense because it needs to load the json file before it does the first checkout (because it's loaded in the main workflow file). :facepalm: so, back to the beginning 8)

Comment (zentol) [reply .github/actions/select_workflow_configs/action.yml:45]: That might be because the checkout action is missing a branch configuration, so it always checks out master.

Comment (XComp) [reply .github/actions/select_workflow_configs/action.yml]: Here's a summary of an offline discussion from yesterday between @zentol and me:
The goal should be to have a per-branch configuration rather than having the configuration being collected in `master` (as it's done in this PR's current proposal). This has multiple advantages:
* The branches' codebase stays independent from each other
* There's a better integration with the GHA UI (no dedicated step to determine the commit hash is necessary)

@zentol came up with a quick POC that involves a [trigger workflow](https://github.com/zentol/flink/actions/workflows/nightly-trigger.yml) in `master` and individual nightly workflow files per branch (e.g. [for master](https://github.com/zentol/flink/actions/workflows/nightly-dev.yml)).

**Review by zentol — APPROVED** [review]
